src/osp/Resource/Resource.h

```diff
@@ -145,11 +145,14 @@ class DependRes
     }
 
     bool empty() const { return m_empty; }
+    operator bool() const { return !empty(); }
     std::string name() const { return m_it->first; }
 
     TYPE_T const& operator*() const { return m_it->second.m_data; }
     TYPE_T& operator*() { return m_it->second.m_data; }
 
+    TYPE_T const* operator->() const { return &m_it->second.m_data; }
```

Comment:
constexpr
noexcept

Reply:
How do I tell when I should be using these? For instance, why do the `operator*()` above not need to be `constexpr`/`noexcept`?

Reply:
They are also good candidates for constexpr and noexcept.

noexcept -> if it can't throw an exception, it should be no except.
constexpr -> if it only does operations that can be conducted at compile time, such as just returning something, then make it constexpr.

Reply:
How do I know it can't throw an exception? For instance, if `m_it` doesn't point to a valid object, wouldn't dereferencing it cause an exception? Since I'm not particularly familiar with exceptions, it's sort of hard to know when one will never be thrown.

Reply:
Whether or not *m_it throws an exception depends on the actual type of m_it.

The real determinant is:

"If this function throws an exception, is there anything meaningful that the program can actually do about it?"

if m_it == std::map::end(), then *m_it is undefined behavior, and you're program *should* crash in order to avoid a security issue. 

So the logical condition that would lead to the situation that might cause an exception to be thrown should, essentially, trigger the program terminating because it strongly implies that your program is in a borked situation anyway.

In that case, the function should be noexcept so that if an exception *is* thrown, the program terminates.

Most functions should be noexcept. They should only ever be not-noexcept if you are actually planning to *catch* the exception thrown from the function. E.g. the function is written with the intention of having a thrown exception *mean* something that can be acted on.